src/Request.php

```diff
@@ -287,27 +287,36 @@ public function parseResponse($data = '', $headersProcessed = false, $returnType
         xml_set_character_data_handler($parser, 'xmlrpc_cd');
         xml_set_default_handler($parser, 'xmlrpc_dh');
 
+        // write xml into a file in order to feed it in chunks to xml_parse
+        $tempXmlFile = tmpfile();
+        $tempXmlFileMetaData = stream_get_meta_data($tempXmlFile);
+        fwrite($tempXmlFile, $data);
```

Comment:
in my opinion you lack `fclose()` here.

at least in C when you write to file and don't close it, and opening the file for read, it can may not see all data as it's buffered in memory and not yet flushed to disk.

this may be different in PHP and in fact that this happens in all same process.

also, php streams api has `php://temp` that has automatic behavior of using memory and fallback to disk if data becames larger:

https://www.php.net/manual/en/wrappers.php.php#wrappers.php.memory:

> php://memory and php://temp are read-write streams that allow temporary data to be stored in a file-like wrapper. The only difference between the two is that php://memory will always store its data in memory, whereas php://temp will use a temporary file once the amount of data stored hits a predefined limit (the default is 2 MB). The location of this temporary file is determined in the same way as the sys_get_temp_dir() function.
>
> The memory limit of php://temp can be controlled by appending /maxmemory:NN, where NN is the maximum amount of data to keep in memory before using a temporary file, in bytes.

also, the whole code seems odd, isn't `$tempXmlFile` already same as `$xmlFile`, why open the handle again?

if you need to reset the stream from beginning use https://php.net/fseek

Reply:
looks like `fclose` on handle created by `tmpfile` deletes the file, so instead `fclose`, you must at least invoke `fflush`, however i suggest to use the streams api instead `php://temp` and `fflush`, `fseek`.

Reply:
@glensc I did some tests yesterday: copying a string into a memory stream does allow us to use stream apis everywhere, but it comes with a cost: memory usage balloons as php does effectively double the data (once for the string, once for the stream). So for the moment I kept usage of strings everywhere, and just added a chunk-and-parse loop around the core xml-parsing call. I did not benchmark that, but my guess is that, worst case, it will add a 4MB memory consumption overhead.
We might want to look into further optimizations in the future, with my current line of thinking being rather to try to parse piecewise while we read from the socket and eschew string usage (ie. avoid reading all data in memory first) but that comes with a price, as it means rewriting the code that deals with http headers, too.